platforms/core-runtime/build-configuration/src/main/java/org/gradle/internal/buildconfiguration/resolvers/DefaultToolchainRepositoriesResolver.java

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.buildconfiguration.resolvers;
+
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.JavaToolchainDownload;
+import org.gradle.jvm.toolchain.JavaToolchainRequest;
+import org.gradle.jvm.toolchain.JavaToolchainResolverRegistry;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
+import org.gradle.jvm.toolchain.JvmImplementation;
+import org.gradle.jvm.toolchain.JvmVendorSpec;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainRequest;
+import org.gradle.jvm.toolchain.internal.DefaultToolchainSpec;
+import org.gradle.jvm.toolchain.internal.JavaToolchainResolverRegistryInternal;
+import org.gradle.jvm.toolchain.internal.RealizedJavaToolchainRepository;
+import org.gradle.platform.BuildPlatform;
+
+import javax.annotation.Nullable;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.gradle.internal.buildconfiguration.resolvers.ToolchainSupportedPlatformsMatrix.getToolchainSupportedBuildPlatforms;
+
+public class DefaultToolchainRepositoriesResolver implements ToolchainRepositoriesResolver {
+
+    private final JavaToolchainResolverRegistryInternal toolchainResolverRegistry;
+    private final ObjectFactory objectFactory;
+
+    public DefaultToolchainRepositoriesResolver(JavaToolchainResolverRegistry toolchainResolverRegistry, ObjectFactory objectFactory) {
+        this.toolchainResolverRegistry = (JavaToolchainResolverRegistryInternal) toolchainResolverRegistry;
+        this.objectFactory = objectFactory;
+    }
+
+    @Override
+    public Map<BuildPlatform, Optional<URI>> resolveToolchainDownloadUrlsByPlatform(
+        JavaLanguageVersion toolchainVersion,
+        @Nullable String toolchainVendor,
+        @Nullable JvmImplementation toolchainImplementation
+    ) throws UnconfiguredToolchainRepositoriesResolver {
+        List<? extends RealizedJavaToolchainRepository> toolchainRepositories = toolchainResolverRegistry.requestedRepositories();
+        if (toolchainRepositories.isEmpty()) {
+            throw new UnconfiguredToolchainRepositoriesResolver();
+        }
+
+        JavaToolchainSpec toolchainSpec = createToolchainSpec(toolchainVersion, toolchainVendor, toolchainImplementation);
+        Map<BuildPlatform, Optional<URI>> toolchainDownloadUrlByPlatformMap = new HashMap<>();
+
+        getToolchainSupportedBuildPlatforms().stream()
+            .map(buildPlatform -> new DefaultJavaToolchainRequest(toolchainSpec, buildPlatform))
+            .forEach(javaToolchainRequest -> {
+                Optional<URI> downloadUrl = resolveToolchainDownloadUrlRequest(toolchainRepositories, javaToolchainRequest);
```

Comment:
The fact that now Gradle allows to specify any string as vendor something that wasn't in the initial proposal , causes to not properly resolving the download URLs. For example in the case of `Bellsoft Liberica` vendor, users must specify `Liberica` as vendor when running `updateDaemonJvm` otherwise the foojay plugin will not be able to resolve URLs properly. This is definitely not desirable since makes more complicate for users to know which vendors to use.

Reply:
Taking `Bellsoft Liberica` as an example, the situtation is even worse.
I downloaded a Java 17 from that vendor and I have it available locally.

Using `Liberica` or `Bellsoft Liberica`, Gradle will not start, saying such a toolchain cannot be found. Running `./gradlew javaToolchains` (after removing the vendor requirement) outputs, amongst others:
```
 + BellSoft Liberica JDK 17.0.13+12-LTS
     | Location:           /Library/Java/JavaVirtualMachines/liberica-jdk-17.jdk/Contents/Home
     | Language Version:   17
     | Vendor:             BellSoft Liberica
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        MacOS java_home

```

So I am digging further into this.

Reply:
This is kind of related to the same and the fact that any string can be used as `vendor`. Basically when parse it if cannot be recognized as [KnownJvmVendor](https://github.com/gradle/gradle/blob/master/platforms/jvm/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/JvmVendor.java#L23C10-L23C24) the specified string will end up being used to find matches for the raw vendor not being desirable since those might not align like in the case of `Liberica` or `Bellsoft Liberica`.

For this reason, the `KnownJvmVendor` should be extended to include missing patters in order to identify better known vendors and handle them properly.

Reply:
Yeah, my digging made me realize the same.

I think the whole system would be easier to use if we had a `Property<JvmVendorSpec>`, similar to the `ToolchainSpec` instead of that String based thing.

I am currently finishing the rebasing work of that PR to test it. Once done, I believe this could be pushed through, but I will want to discuss changing the task configuration.

Reply:
Absolutely agree, enforcing users to specify `JvmVendorSpec` will simplify things considering that most projects will use foojay plugin instead some custom resolver plugin. The IDE itself will add Foojay plugin by default for new projects but also when migrating existing ones. That's why there are going to be just few cases when projects will have custom resolver plugins applied for this reason make also sense to have Foojay plugin embedded in the Gradle itself so this will simplify the requirements for projects. 

In addition, its also useful for users in case they specify any unknown vendors to have clear message about supported ones. 
```
Value 'unknown-vendor' given for toolchainVendor is an invalid Java vendor.  Possible values are [ADOPTIUM, ADOPTOPENJDK, AMAZON, APPLE, AZUL, BELLSOFT, GRAAL_VM, HEWLETT_PACKARD, IBM, JETBRAINS, MICROSOFT, ORACLE, SAP, TENCENT]
```

For sure this can be something to revisit in the future if users requests additional support for custom vendors for Daemon toolchain